### PR TITLE
Very Basic FPGA flow with flow tests

### PR DESCRIPTION
**Pull request by petergrossmann21**

### CODE REVIEW

**Comment by gadfort** [comment]
@petergrossmann21 please merge with the current main branch and fix the pure python tests

File: siliconcompiler/flows/fpgaflow.py [hunk 1]
```diff
@@ -103,108 +98,6 @@ def setup(chip, flowname='fpgaflow'):
     return flow
 
 
-##################################################
-def flow_lookup(partname):
```
Comment (aolofsson) [inline siliconcompiler/flows/fpgaflow.py:107]: Why was the flow lookup deleted?

File: siliconcompiler/tools/yosys/syn_fpga.tcl [hunk 2]
```diff
@@ -4,19 +4,159 @@ set job_name [dict get $sc_cfg option jobname]
...
+#          As of 1/13/2023, what's here is a best-effort first stab at getting
+#          all of the commands exercised in a first-order fashion and defining
+#          part number strings that could be used to specify each supported
+#          part number.  Where available, part number prefixes have been gleaned
+#          from DigiKey so that there is a correspondence here between
+#          real product part numbers and the if/else statement selecting a
+#          synthesis command.  It will be necessary to do some vetting and
+#          decide what to really support and what to omit based on the yosys
+#          support (or the place and route support) being too skimpy.
+#          -PG 1/13/2023
```
Comment (aolofsson) [inline siliconcompiler/tools/yosys/syn_fpga.tcl:36]: We should be consistent in our code base. afaik there are no other examples of file authors or blocks/signatures, so these should be fixed.  (ie no need for PG 1/13/2023)

**Review by aolofsson — CHANGES_REQUESTED** [review]

**Comment by petergrossmann21** [comment]
Code has been revised accordingly.

File: siliconcompiler/flows/fpgaflow.py (hunk 1 above)
Comment (petergrossmann21) [reply siliconcompiler/flows/fpgaflow.py:107]: It has essentially been replaced by the if/else logic setting the flowpipe variable.  This allows things to be a little more a la carte between flows and targets.  There's an open philosophical debate to have about separation of concerns for FPGA flows between flows and targets that the current code largely defers by only supporting open source flows for VPR and nextpnr and only for specific demo architectures (ice40 and my academic example for VPR).  The priority has been given to supporting a subset of what was originally envisioned that can be tested in CI, rather than including code that cannot be tested due to vendor-specific SW requirements.  Followup dicussions recommended to scope work beyond the reach of this PR.

File: siliconcompiler/tools/genfasm/bitstream.py [hunk 3]
```diff
@@ -37,26 +32,27 @@ def setup(chip):
...
-    tool = "genfasm"
-
-    chip.add('tool', tool, 'task', task, 'option', [f"--route_chan_width {find_chann_width()}"],
-             step=step, index=index)
+# def pre_process(chip):
 
+    # step = chip.get('arg', 'step')
+    # index = chip.get('arg', 'index')
+    # task = chip._get_task(step, index)
+    # tool = "genfasm"
```
Comment (gadfort) [inline siliconcompiler/tools/genfasm/bitstream.py:40]: remove

File: siliconcompiler/tools/genfasm/bitstream.py [hunk 4]
```diff
@@ -37,26 +32,27 @@ def setup(chip):
...
-    return -1
+
+
+def post_process(chip):
+    ''' Tool specific function to run after step execution
+    '''
+
+    # step = chip.get('arg', 'step')
+    # index = chip.get('arg', 'index')
+    # task = chip._get_task(step, index)
```
Comment (gadfort) [inline siliconcompiler/tools/genfasm/bitstream.py:53]: remove

File: siliconcompiler/tools/vpr/route.py [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+    Perform automated place and route with VPR
+    '''
+
+    tool = 'vpr'
+    step = chip.get('arg', 'step')
+    index = chip.get('arg', 'index')
+    task = chip._get_task(step, index)
+
+    chip.set('tool', tool, 'exe', tool, clobber=False)
+    chip.set('tool', tool, 'version', '0.0', clobber=False)
```
Comment (gadfort) [inline siliconcompiler/tools/vpr/route.py:19]: don't set if we don't have a way to check versions
If there is a method to check the VPR version, that should be implemented.

File: siliconcompiler/tools/vpr/route.py [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+def pre_process(chip):
+
+    # have to rename the net connected to unhooked pins from $undef to unconn
+    # as VPR uses unconn keywords to identify unconnected pins
+
+    step = chip.get('arg', 'step')
+    index = chip.get('arg', 'index')
+    design = chip.top()
+    blif_file = f"{chip._getworkdir()}/{step}/{index}/inputs/{design}.blif"
+    print(blif_file)
```
Comment (gadfort) [inline siliconcompiler/tools/vpr/route.py:55]: remove print.

File: siliconcompiler/tools/vpr/route.py [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+    with open(blif_file, 'r+') as f:
+        netlist = f.read()
+        f.seek(0)
+        netlist = re.sub(r'\$undef', 'unconn', netlist)
+        f.write(netlist)
+        f.truncate()
```
Comment (gadfort) [inline siliconcompiler/tools/vpr/route.py:61]: why are you rewriting the input file? If this is needed, you should instead rename the input file something like `sc_{design}.blif` since we use hardlinks for the input files, which means you are actually overwriting the output of the previous step. and up the flow to point to the `sc_{}.blif` file instead.

File: siliconcompiler/tools/vpr/route.py [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+    chip.set('tool', tool, 'exe', tool, clobber=False)
+    chip.set('tool', tool, 'version', '0.0', clobber=False)
+
+    chip.set('tool', tool, 'task', task, 'threads', os.cpu_count(),
+             step=step, index=index, clobber=False)
+
+    # TO-DO: PRIOROTIZE the post-routing packing results?
+    design = chip.top()
+    chip.add('tool', tool, 'task', task, 'output', design + '.route', step=step, index=index)
+    chip.add('tool', tool, 'task', task, 'output', 'vpr_stdout.log', step=step, index=index)
```
Comment (gadfort) [inline siliconcompiler/tools/vpr/route.py:27]: is this file an actual output? if not, it should not be recorded as an output

File: siliconcompiler/tools/vpr/vpr.py [hunk 9]
```diff
@@ -26,6 +26,47 @@ def make_docs(chip):
...
+
+    topmodule = chip.top()
+    blif = f"inputs/{topmodule}.blif"
+
+    for arch in chip.get('fpga', 'arch'):
+        options.append(arch)
+
+    options.append(blif)
+
+    if 'sdc' in chip.getkeys('input'):
```
Comment (gadfort) [inline siliconcompiler/tools/vpr/vpr.py:45]: input/constraints/sdc or xdc? and why put it in FPGA vs. constraint?

File: siliconcompiler/tools/vpr/vpr.py [hunk 10]
```diff
@@ -26,6 +26,47 @@ def make_docs(chip):
...
+
+    # ***NOTE: For real FPGA chips you need to specify the routing channel
+    #          width explicitly.  VPR requires an explicit routing channel
+    #          with when --read_rr_graph is used (typically the case for
+    #          real chips).  Otherwise VPR performs a binary search for
+    #          the minimum routing channel width that the circuit fits in.
+    # Given the above, it may be appropriate to couple these variables somehow,
+    # but --route_chan_width CAN be used by itself.
+    num_routing_channels = chip.get('tool', 'vpr', 'task', 'apr', 'var', 'route_chan_width',
+                                    step=step, index=index)
```
Comment (gadfort) [inline siliconcompiler/tools/vpr/vpr.py:63]: this gets called on all tasks, please set this correctly for each task. Using `tool, task = chip._get_tool_task(step, index)`

File: siliconcompiler/tools/yosys/syn_fpga.tcl [hunk 11]
```diff
@@ -4,19 +4,157 @@ set job_name [dict get $sc_cfg option jobname]
 set step [dict get $sc_cfg arg step]
 set index [dict get $sc_cfg arg index]
 
-#TODO: add logic that remaps yosys built in name based on part number
+set sc_memmap_techlib "None"
+set sc_mem_techlib "None"
+set sc_lut_techlib "None"
+set sc_dsp_techlib "None"
+set sc_flop_techlib "None"
```
Comment (gadfort) [inline siliconcompiler/tools/yosys/syn_fpga.tcl:11]: what are these used for?

File: siliconcompiler/targets/fpga_vpr_flow_demo.py [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+    utils.set_common_showtools(chip)
+
+    # 3. Select default flow
+    chip.set('option', 'mode', 'fpga', clobber=False)
+    chip.set('option', 'flow', 'fpgaflow', clobber=False)
+
+    # 4. Target-specific synthesis settings
+    chip.add('tool', 'yosys', 'task', 'syn', 'var', 'lut_size', '4')
+    chip.add('tool', 'yosys', 'task', 'syn', 'var', 'memmap', 'None')
+    chip.add('tool', 'yosys', 'task', 'syn', 'var', 'techmap', 'None')
```
Comment (gadfort) [inline siliconcompiler/targets/fpga_vpr_flow_demo.py:28]: where are these used? ans if they are files they should be in [..., 'file', 'techmap']

File: siliconcompiler/tools/vpr/place.py [hunk 13]
```diff
@@ -19,24 +21,15 @@ def setup(chip):
     chip.set('tool', tool, 'task', task, 'threads', os.cpu_count(),
              step=step, index=index, clobber=False)
 
-    # TO-DO: PRIOROTIZE the post-routing packing results?
     design = chip.top()
     chip.set('tool', tool, 'task', task, 'output', design + '.net', step=step, index=index)
     chip.add('tool', tool, 'task', task, 'output', design + '.place', step=step, index=index)
-    chip.add('tool', tool, 'task', task, 'output', design + '.route', step=step, index=index)
     chip.add('tool', tool, 'task', task, 'output', 'vpr_stdout.log', step=step, index=index)
```
Comment (gadfort) [inline siliconcompiler/tools/vpr/place.py:27]: not an output?

File: siliconcompiler/tools/vpr/vpr.py [hunk 14]
```diff
@@ -26,6 +26,47 @@ def make_docs(chip):
...
+
+    options.append(blif)
+
+    if 'sdc' in chip.getkeys('input'):
+        options.append(f"--sdc_file {chip.get('input', 'fpga', 'sdc', step=step, index=index)}")
+    else:
+        options.append("--timing_analysis off")
+
+    # Routing graph XML:
+    rr_graphs = chip.get('tool', 'vpr', 'task', 'apr', 'var', 'rr_graph', step=step, index=index)
```
Comment (gadfort) [inline siliconcompiler/tools/vpr/vpr.py:51]: document what this option is with chip.get(',,,,', <helpstring>, field='help')

File: siliconcompiler/tools/vpr/vpr.py [hunk 15]
```diff
@@ -26,6 +26,47 @@ def make_docs(chip):
...
+        options.append("--read_rr_graph " + rr_graphs[0])
+
+    # ***NOTE: For real FPGA chips you need to specify the routing channel
+    #          width explicitly.  VPR requires an explicit routing channel
+    #          with when --read_rr_graph is used (typically the case for
+    #          real chips).  Otherwise VPR performs a binary search for
+    #          the minimum routing channel width that the circuit fits in.
+    # Given the above, it may be appropriate to couple these variables somehow,
+    # but --route_chan_width CAN be used by itself.
+    num_routing_channels = chip.get('tool', 'vpr', 'task', 'apr', 'var', 'route_chan_width',
```
Comment (gadfort) [inline siliconcompiler/tools/vpr/vpr.py:62]: document option

File: siliconcompiler/tools/yosys/syn_fpga.tcl [hunk 16]
```diff
@@ -4,19 +4,157 @@ set job_name [dict get $sc_cfg option jobname]
...
+    yosys share
+    yosys opt
+    yosys memory -nomap
+    yosys opt -full
+
+    #Do our own thing from here
+    yosys techmap -map +/techmap.v
+
+    #***TO DO:  Figure out how to deal with this in an SC-compatible way
+    if { $sc_flop_techlib eq "None" } {
```
Comment (gadfort) [inline siliconcompiler/tools/yosys/syn_fpga.tcl:83]: if this is a file, then:
```
if { $sc_flop_techlib == "" } {
} else {
....?
}
```

File: siliconcompiler/tools/yosys/syn_fpga.tcl [hunk 17]
```diff
@@ -4,19 +4,157 @@ set job_name [dict get $sc_cfg option jobname]
...
+
+    #opt
+    #opt_clean
+
+    #Perform preliminary buffer insertion before passing to ABC to help reduce
+    #the overhead of final buffer insertion downstream
+    yosys insbuf
+
+    yosys abc -lut $sc_syn_lut_size -dff
+    #abc -script $env(FPGA_ARCHITECT_ROOT)/test/scripts/fpgaa_abc_flow.scr
```
Comment (gadfort) [inline siliconcompiler/tools/yosys/syn_fpga.tcl:95]: remove

File: siliconcompiler/tools/yosys/syn_fpga.tcl [hunk 18]
```diff
@@ -4,19 +4,157 @@ set job_name [dict get $sc_cfg option jobname]
...
+    #Do our own thing from here
+    yosys techmap -map +/techmap.v
+
+    #***TO DO:  Figure out how to deal with this in an SC-compatible way
+    if { $sc_flop_techlib eq "None" } {
+	yosys dfflegalize -cell \$_DFF_P_ 01 -cell \$_DFF_P??_ 01 
+    }
+
+    #opt
+    #opt_clean
```
Comment (gadfort) [inline siliconcompiler/tools/yosys/syn_fpga.tcl:88]: remove

File: siliconcompiler/tools/yosys/syn_fpga.tcl [hunk 19]
```diff
@@ -4,19 +4,157 @@ set job_name [dict get $sc_cfg option jobname]
...
+    yosys abc -lut $sc_syn_lut_size -dff
+    #abc -script $env(FPGA_ARCHITECT_ROOT)/test/scripts/fpgaa_abc_flow.scr
+
+    #***TO DO:  Wrap this in a verbosity flag
+    #yosys write_verilog -noexpr -selected ${sc_design}_abc.v
+
+    #Clean up the design before buffer insertion to avoid blowing up the design
+    #size with that step
+    #opt
+    #clean
```
Comment (gadfort) [inline siliconcompiler/tools/yosys/syn_fpga.tcl:103]: remove

File: siliconcompiler/tools/yosys/syn_fpga.tcl [hunk 20]
```diff
@@ -4,19 +4,157 @@ set job_name [dict get $sc_cfg option jobname]
...
+	#insbuf -buf $buffer_name $buf_input_name $buf_output_name
+    #}
+
+    #***NOTE:  hilomap needs to happen before tech mapping so that
+    #          we can run the hierarchy check and prevent the tiecell
+    #          library from being treated like part of the design
+
+    #if { $tiecells == 1 } {
+	#hilomap -hicell $tiehi_cell $tiehi_output_name -locell $tielo_cell $tielo_output_name
+    #}
```
Comment (gadfort) [inline siliconcompiler/tools/yosys/syn_fpga.tcl:129]: why are these commented out?

File: siliconcompiler/tools/yosys/syn_fpga.tcl [hunk 21]
```diff
@@ -4,19 +4,157 @@ set job_name [dict get $sc_cfg option jobname]
...
+    ########################################################
+    # Technology Mapping
+    ########################################################
+    if [dict exists $sc_cfg tool $sc_tool var $sc_step $sc_index techmap ] {
+	set sc_techmap [ dict get $sc_cfg tool $sc_tool var $sc_step $sc_index techmap ]
+	# techmap_defines = [ dict get $sc_cfg tool $sc_tool var $sc_step $sc_index techmap_define ]
+	# techmap_define_args = ''
+	# foreach techmap_define $techmap_defines {
+	#    set techmap_define_args "$techmap_define_args -D$techmap_define"
+	# }
```
Comment (gadfort) [inline siliconcompiler/tools/yosys/syn_fpga.tcl:140]: remove

File: siliconcompiler/tools/yosys/syn_fpga.tcl [hunk 22]
```diff
@@ -4,19 +4,157 @@ set job_name [dict get $sc_cfg option jobname]
...
+	foreach mapfile $sc_techmap {
+	    # Add $techmap_define_args when ready
+	    yosys techmap -map $mapfile
```
Comment (gadfort) [inline siliconcompiler/tools/yosys/syn_fpga.tcl:143]: In the asic flow we need to follow a techmap with:
https://github.com/siliconcompiler/siliconcompiler/blob/cebeb2f78bf4fc6498adba06aac80dcb702190b9/siliconcompiler/tools/yosys/syn_asic.tcl#L183C1-L188C2
In case this adds some unmapped logic like `assign x = ~b`

File: tests/core/test_target.py [hunk 23]
```diff
@@ -13,11 +13,11 @@ def test_target_valid():
...
 
 def test_target_fpga_valid():
-    '''Ensure that the FPGA flow allows legal part names and sets mode
+    '''Ensure that the VPR FPGA flow allows legal part names and sets mode
     correctly.'''
     chip = siliconcompiler.Chip('test')
-    chip.set('fpga', 'partname', 'ice40')
-    chip.load_target("fpgaflow_demo")
+    chip.set('fpga', 'partname', 'example_arch_X005Y005')
+    chip.load_target('fpga_vpr_flow_demo')
```
Comment (gadfort) [inline tests/core/test_target.py:20]: use `from siliconcompiler.targets import fpga_vpr_flow_demo`

File: tests/flows/test_fpgaflow.py [hunk 24]
```diff
@@ -0,0 +1,65 @@
...
+
+    # 4.  VPR Setup
+    xml_file = os.path.join(arch_root, f'{arch_name}.xml')
+    rr_graph_file = os.path.join(arch_root, f'{arch_name}_rr_graph.xml')
+
+    chip.set('fpga', 'arch', xml_file)
+    # ***NOTE:  If the RR graph is not specified, the FASM bitstream will
+    #           generate but omit any bitstream data for programmable
+    #           interconnect (SBs and CBs); meaning that the FPGA will
+    #           not be correctly programmed. -PG 5/17/2023
```
Comment (gadfort) [inline tests/flows/test_fpgaflow.py:48]: more `PG <date>` remove

File: tests/flows/test_fpgaflow.py [hunk 25]
```diff
@@ -0,0 +1,65 @@
...
+    # 5. Load target
+    chip.load_target(fpga_vpr_flow_demo)
+
+    # 6. Set flow
+
+    chip.run()
+
+    fasm_file = chip.find_result('fasm', step='bitstream')
+
+    assert fasm_file.endswith(f'{top_module}.fasm')
```
Comment (gadfort) [inline tests/flows/test_fpgaflow.py:61]: assert this file exists

File: tests/flows/test_fpgaflow_apr.py [hunk 26]
```diff
@@ -0,0 +1,56 @@
...
+    # 5. Load target
+    chip.load_target(fpga_vpr_flow_demo)
+
+    # 6. Set flow
+
+    chip.run()
+
+    route_file = chip.find_result('route', step='route')
+
+    assert route_file.endswith(f'{top_module}.route')
```
Comment (gadfort) [inline tests/flows/test_fpgaflow_apr.py:52]: assert this file exists

**Review by gadfort — CHANGES_REQUESTED** [review]
Over all looks good, please address the comments

File: siliconcompiler/flows/fpgaflow.py (hunk 1 above)
Comment (aolofsson) [reply siliconcompiler/flows/fpgaflow.py:107]: This seems to break the existing functionality? There is a reason why I made the flow somwehat complex. I wanted to be able to use one entry point, with the part name dicating what path to take. Of course we can always specify separate flows and targets, but this is a bigger mental load.

Comment (petergrossmann21) [reply siliconcompiler/flows/fpgaflow.py:107]: In some cases, breaking existing functionality was intentional.  To points made by @gadfort when the two of us were comparing notes, there are limits to what SC can support w/respect to commercial FPGA RTL to bitstream flows because dependencies are introduced on tools (e.g. Vivado) that come with terms of use that limit their open adoption.  Additionally, the Venn diagram of tools supported by yosys + (VPR+genfasm || nextpnr) is basically ICE40 and (maybe?) Xilinx 7 series.  I wanted to solve the minimum viable problem that put a new flow back on SC main with minimal development effort, even if it meant a temporary step backwards that pushed us to create new PRs to re-introduce things that worked at point (notably, some limited support for Xilinx).  

As for the "one entry point set by part number" vs. a more "build your flow/target pair for a given FPGA family / CAD flow pair" approach, I'm not 100% married to anything.  Part of how I arrived at the current approach was simply trying to develop a mental model for what it meant to be a flow and what it meant to be a target.  In the ASIC arena, SC only has one tool chain to worry about (OpenROAD and friends) and many process technologies + associated cell libraries.  For the FPGA case, there are two open source tool chains, and (theoretically) many families of parts to program.  Following the analogy that flow == ASIC tool chain == FPGA tool chain and target == ASIC process == FPGA parts family, I arrived at what you see here.

File: siliconcompiler/tools/yosys/syn_fpga.tcl (hunk 2 above)
Comment (petergrossmann21) [reply siliconcompiler/tools/yosys/syn_fpga.tcl]: Done

File: siliconcompiler/tools/genfasm/bitstream.py (hunk 3 above)
Comment (petergrossmann21) [reply siliconcompiler/tools/genfasm/bitstream.py]: Done.

File: siliconcompiler/tools/genfasm/bitstream.py (hunk 4 above)
Comment (petergrossmann21) [reply siliconcompiler/tools/genfasm/bitstream.py]: Done.

File: siliconcompiler/tools/yosys/syn_fpga.tcl (hunk 22 above)
Comment (petergrossmann21) [reply siliconcompiler/tools/yosys/syn_fpga.tcl:143]: Based on empirical evidence, the common FPGA case seems to be that abc mapping to LUTs addresses such issues.  I'm a bit concerned that there are corner cases I've overlooked where the ASIC flow analogy arises, but would need to run the flow on a large number of test cases to try and find a failure.

File: tests/flows/test_fpgaflow_apr.py (hunk 26 above)
Comment (petergrossmann21) [reply tests/flows/test_fpgaflow_apr.py]: Done

File: tests/flows/test_fpgaflow.py (hunk 24 above)
Comment (petergrossmann21) [reply tests/flows/test_fpgaflow.py]: Done

File: siliconcompiler/tools/vpr/place.py (hunk 13 above)
Comment (petergrossmann21) [reply siliconcompiler/tools/vpr/place.py:27]: Correct, if you run the placement step by itself, as is the case here, it does not produce a .route file.

File: siliconcompiler/tools/yosys/syn_fpga.tcl (hunk 11 above)
Comment (petergrossmann21) [reply siliconcompiler/tools/yosys/syn_fpga.tcl:11]: These are placeholders that are temporarily disabling the flow's ability to perform technology mapping to user-defined tech libraries for each of the major hard block types in modern FPGAs.  Based on other feedback I've received on this PR, I'm guessing you're going to want me to strip down the syn_fpga.tcl and my testcases to eliminate all of this until I have a test case that performs technology mapping?

File: siliconcompiler/targets/fpga_vpr_flow_demo.py (hunk 12 above)
Comment (petergrossmann21) [reply siliconcompiler/targets/fpga_vpr_flow_demo.py:28]: Noted, will adjust based on overall feedback regarding tech mapping support.

File: siliconcompiler/tools/vpr/route.py (hunk 8 above)
Comment (petergrossmann21) [reply siliconcompiler/tools/vpr/route.py:27]: Can't remember why I included the log file in the list of outputs; removed.

File: siliconcompiler/tools/vpr/route.py (hunk 7 above)
Comment (petergrossmann21) [reply siliconcompiler/tools/vpr/route.py:61]: This was inherited from previous development efforts.  Removing it.

File: siliconcompiler/tools/vpr/vpr.py (hunk 9 above)
Comment (petergrossmann21) [reply siliconcompiler/tools/vpr/vpr.py:45]: This is inherited from previous developments.  All that needs to happen here is that when an SDC file is provided, the command line option to pass the SDC file is added to the VPR command line, and when no SDC is provided, the option is omitted and we run without one.

File: siliconcompiler/tools/yosys/syn_fpga.tcl (hunk 11 above)
Comment (petergrossmann21) [reply siliconcompiler/tools/yosys/syn_fpga.tcl]: Per above comment, I've implemented the reduced syn_fpga.tcl.  We can add things back at such as a time as I am able to furnish an architecture with hard macros that require technology mapping.

File: siliconcompiler/flows/fpgaflow.py (hunk 1 above)
Comment (aolofsson) [reply siliconcompiler/flows/fpgaflow.py:107]: Got it, thanks for describing the process. My view is that we should have as few entry points as possible to reduce the mental load on the user to memorize arbitrary names. In the case of FPGAs, whether using vpr, vivado, quartus we always have to specify which partname we are targeting.

I feel strongly that we should simplify things for the user as much as possible and it is my view that this is done by limiting the introuction of new things. This is why the fpgaflow. Can you restore the previous logic while enabling the new target?

**Comment by gadfort** [comment]
@aolofsson this is identical to the previous PR. No additional schema changes and contains the same updates to the fpga flow to restore the previous flow selection.

**Review by gadfort — APPROVED** [review]

**Review by aolofsson — APPROVED** [review]
LGTM, sorry for the delay
